### PR TITLE
Use MonadRec for many

### LINKED ISSUE #9: Not stack safe
A simple parser like `many (string "a")` will fail on an input of size 3000.

Due to lack of general TCO - not sure if it's inherent in the continuation based design of the `Parser` type or just the mutual recursion of things like `many`/`many1` and similar.

### PR DESCRIPTION
**Pull request by garyb**
Rebased version of #21 (thanks @nwolverson!)

Resolves #21, resolves #9.

### CODE REVIEW

**Comment by paf31** [comment]
Looks great, but should this go in `lists` instead?

File: src/Text/Parsing/StringParser/Combinators.purs [hunk 1]
```diff
@@ -22,14 +24,16 @@ lookAhead (Parser p) = Parser \s ->
 
 -- | Match zero or more times.
 many :: forall a. Parser a -> Parser (List a)
-many p = many1 p <|> pure Nil
+many p = tailRecM go Nil
+  where
+  go :: List a -> Parser (Step (List a) (List a))
+  go acc = do
+    aa <- (Loop <$> p) <|> pure (Done unit)
+    pure $ bimap (flip Cons acc) (\_ -> reverse acc) aa
```
Comment (paf31) [inline src/Text/Parsing/StringParser/Combinators.purs]: or `_ : acc`?

**Comment by garyb** [comment]
Oh right, we already have it there. In that case I'll drop it and add a re-export instead.

**Comment by paf31** [comment]
True, but the version in `lists` isn't stack safe.

**Comment by garyb** [comment]
How's this?

**Comment by paf31** [comment]
Looks great!

**Comment by nwolverson** [comment]
Nice :)

**Pull request merged**